### PR TITLE
AX: input[type=date] individual fields are announced as "group"

### PR DESCRIPTION
**Pull request by Christop406**
#### d4f4aedd06563422f36fff424b3592646262e0f3
<pre>
AX: input[type=date] individual fields are announced as &quot;group&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=251544">https://bugs.webkit.org/show_bug.cgi?id=251544</a>
rdar://problem/104928713

Reviewed by Aditya Keerthi.

This fixes the bug by adding labels representing the &quot;day&quot;, &quot;month&quot;, and &quot;year&quot; inputs so that a screen reader has something to read,
instead of just &quot;group&quot; - it also switches the &quot;role&quot; attribute of the same divs to &quot;spinbutton&quot; to more accurately
reflect how they are used.

* LayoutTests/accessibility/date-input-aria-label-and-value-expected.txt: Added.
* LayoutTests/accessibility/date-input-aria-label-and-value.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
(WebCore::DateTimeDayFieldElement::create):
(WebCore::DateTimeMonthFieldElement::create):
(WebCore::DateTimeSymbolicMonthFieldElement::create):
(WebCore::DateTimeYearFieldElement::create):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
(WebCore::DateTimeNumericFieldElement::setEmptyValue):
(WebCore::DateTimeNumericFieldElement::setValueAsInteger):
(WebCore::DateTimeNumericFieldElement::setARIAValueAttributesWithInteger):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXDateFieldMonthText):
(WebCore::AXDateFieldDayText):
(WebCore::AXDateFieldYearText):
* Source/WebCore/platform/LocalizedStrings.h:

Canonical link: <a href="https://commits.webkit.org/261123@main">https://commits.webkit.org/261123@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/62ae5663fb8f8f8c405f63b5fb89bbe259d231bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110578 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1934 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119446 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10790 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102874 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15703 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98884 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43995 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97663 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30551 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85909 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12311 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31889 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12873 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8853 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18237 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51505 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7707 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14755 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by Christop406** [comment]
Adding to this - I added a couple new translation strings as part of this PR - if this was not the correct way to go about this, please let me know by all means. I didn't add translations for them because I wasn't sure if there was a process in place for this but would be happy to help get them.

File: Source/WebCore/en.lproj/Localizable.strings [hunk 1]
```diff
@@ -1911,4 +1920,3 @@
 
 /* Option in segmented control for choosing list type in text editing */
 "•" = "•";
-
```
Comment (pxlcoder) [inline Source/WebCore/en.lproj/Localizable.strings]: Please undo this change.

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp [hunk 2]
```diff
@@ -219,7 +223,10 @@ DateTimeMonthFieldElement::DateTimeMonthFieldElement(Document& document, FieldOw
 
 Ref<DateTimeMonthFieldElement> DateTimeMonthFieldElement::create(Document& document, FieldOwner& fieldOwner)
 {
-    auto element = adoptRef(*new DateTimeMonthFieldElement(document, fieldOwner));
+    auto dateEl = new DateTimeMonthFieldElement(document, fieldOwner);
```
Comment (shivamidow) [inline Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: /* Unofficial Review */
dateEl -> dateElement? https://webkit.org/code-style-guidelines/#names-basic

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp [hunk 3]
```diff
@@ -219,7 +223,10 @@ DateTimeMonthFieldElement::DateTimeMonthFieldElement(Document& document, FieldOw
...
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldMonthText() });
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
+    auto element = adoptRef(*dateEl);
```
Comment (shivamidow) [inline Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: /* Unofficial Review */
Is there a reason why we need dateEl here? Simply using element seems fine.
element->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString...);

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp [hunk 4]
```diff
@@ -49,6 +51,8 @@ Ref<DateTimeDayFieldElement> DateTimeDayFieldElement::create(Document& document,
     static MainThreadNeverDestroyed<const AtomString> dayPseudoId("-webkit-datetime-edit-day-field"_s);
     ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
     element->initialize(dayPseudoId);
+    element->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
```
Comment (pxlcoder) [inline Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: Can we get the "spinbutton" string using `accessibilityRoleToString` from `AccessibilityObjectInterface.h`?

File: Source/WebCore/platform/LocalizedStrings.cpp [hunk 5]
```diff
@@ -719,6 +719,21 @@ String AXTimeFieldText()
     return WEB_UI_STRING("time field", "accessibility role description for a time field.");
 }
 
+String AXDateFieldMonthText()
+{
+    return WEB_UI_STRING("month", "accessibility label for a date field month input.");
```
Comment (shivamidow) [inline Source/WebCore/platform/LocalizedStrings.cpp:713]: /* Unofficial Review */
I suggest you to match these strings with others like "accessibility role description for XXX."

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp [hunk 6]
```diff
@@ -219,7 +223,10 @@ DateTimeMonthFieldElement::DateTimeMonthFieldElement(Document& document, FieldOw
 
 Ref<DateTimeMonthFieldElement> DateTimeMonthFieldElement::create(Document& document, FieldOwner& fieldOwner)
 {
-    auto element = adoptRef(*new DateTimeMonthFieldElement(document, fieldOwner));
+    auto dateEl = new DateTimeMonthFieldElement(document, fieldOwner);
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldMonthText() });
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
```
Comment (pxlcoder) [inline Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: Can you do this after the call to `initialize` below?

And undo the change to the variable name.

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp [hunk 7]
```diff
@@ -278,6 +285,8 @@ Ref<DateTimeSymbolicMonthFieldElement> DateTimeSymbolicMonthFieldElement::create
     static MainThreadNeverDestroyed<const AtomString> monthPseudoId("-webkit-datetime-edit-month-field"_s);
     ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
     element->initialize(monthPseudoId);
+    element->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldMonthText() });
+    element->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
```
Comment (pxlcoder) [inline Source/WebCore/html/shadow/DateTimeFieldElements.cpp:288]: Let's be consistent with the order we set this in. Either label, then role, or vice versa.

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp [hunk 8]
```diff
@@ -117,19 +118,27 @@ void DateTimeNumericFieldElement::setEmptyValue(EventBehavior eventBehavior)
...
 }
 
 void DateTimeNumericFieldElement::setValueAsIntegerByStepping(int value)
 {
     m_typeAheadBuffer.clear();
     setValueAsInteger(value, DispatchInputAndChangeEvents);
+    setValueText(value);
+}
+
+void DateTimeNumericFieldElement::setValueText(int value)
```
Comment (pxlcoder) [inline Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp]: I would rename this to `setARIAValueTextWithInteger` to be more explicit.

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.h [hunk 9]
```diff
@@ -59,6 +59,7 @@ class DateTimeNumericFieldElement : public DateTimeFieldElement {
     void setValueAsInteger(int, EventBehavior = DispatchNoEvent) final;
     void stepDown() final;
     void stepUp() final;
+    void setValueText(int);
```
Comment (shivamidow) [inline Source/WebCore/html/shadow/DateTimeNumericFieldElement.h]: /* Unofficial Review */
This method is confusing. We set a text value but take int? Maybe, we can change it to something like setValueText(String&).

File: Source/WebCore/platform/LocalizedStrings.cpp [hunk 10]
```diff
@@ -719,6 +719,21 @@ String AXTimeFieldText()
...
+}
+
+String AXDateFieldDayText()
+{
+    return WEB_UI_STRING("day", "accessibility label for a date field month input.");
```
Comment (twilco) [inline Source/WebCore/platform/LocalizedStrings.cpp]: Seems like a copy-paste error here —  "month input" should probably be "day input". Same issue for `AXDateFieldYearText()`.

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp [hunk 11]
```diff
@@ -117,19 +118,27 @@ void DateTimeNumericFieldElement::setEmptyValue(EventBehavior eventBehavior)
...
+{
+    setAttributeWithoutSynchronization(HTMLNames::aria_valuetextAttr, AtomString { WTF::String::number(value) });
```
Comment (twilco) [inline Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp]: I think you can use `String::number(value)` rather than `AtomString { WTF::String::number(value) }` here.

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp (hunk 4 above)
Comment (twilco) [reply Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: I think `accessibilityRoleToString` may not be what we want, as that function is more for debugging and exposes things that don't actually map to real ARIA roles.

We do have a mapping from WebCore roles to ARIA roles via `AccessibilityObject::reverseAriaRoleMap()`, but that's not exposed outside the class right now. In my opinion hardcoding "spinbutton" is OK here, but if you feel we should do this as a lookup, then the right way to do it would be exposing a public static function (i.e. `AccessibilityObject::webCoreRoleToAriaRole(AccessibilityRole)`) that uses this reverse role map.

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp [hunk 12]
```diff
@@ -117,19 +118,27 @@ void DateTimeNumericFieldElement::setEmptyValue(EventBehavior eventBehavior)
...
     m_hasValue = true;
     updateVisibleValue(eventBehavior);
+    setValueText(value);
 }
 
 void DateTimeNumericFieldElement::setValueAsIntegerByStepping(int value)
 {
     m_typeAheadBuffer.clear();
     setValueAsInteger(value, DispatchInputAndChangeEvents);
+    setValueText(value);
```
Comment (twilco) [inline Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp]: Is this call to `setValueText` necessary? The line above calls `setValueAsInteger`, which calls `setValueText`, which I believe makes the call on this line redundant.

**Review by twilco — COMMENTED** [review]
Thanks for working on this!

It would be nice to have a layout test to ensure future changes don't break accessibility. Two cases come to mind:

1. Put keyboard focus onto a date field and send up / down keys, and confirm the accessibility label is what we expect
2. Update the date input's value via JS, and again verify the accessibility label of the input as a whole and the sub-fields within

You can read about layout tests here:

https://github.com/WebKit/WebKit/blob/main/Introduction.md#running-layout-tests

You'll want to add your new test to `LayoutTests/accessibility`.

Within the context of layout tests, you can use a global `eventSender` object to simulate events. So for the first testcase, you may want something like this:

```
// Double childAtIndex to get through `<div pseudo="-webkit-datetime-edit">` to `<div pseudo="-webkit-datetime-edit-fields-wrapper">`
var axDateFieldWrapper = accessibilityController.accessibleElementById("date-input").childAtIndex(0).childAtIndex(0);

document.getElementById("date-input").focus();
// Focus should be on the first of the three groups (e.g. the "month" field)

eventSender.keyDown("upArrow");
// Verify the description. aria-valuetext is exposed through AccessibilityNodeObject::valueDescription(), which is in turn exposed through AccessibilityUIElement::valueDescription() for layout tests.
testOutput += expect("axDateFieldWrapper.childAtIndex(0).valueDescription", "AXValueDescription: ...some expected value...");

// Maybe try some more up / down arrow keys, and verify appropriately

// Tab to the next field and repeat
eventSender.keyDown("\t");
...
```

In terms of general test structure, https://github.com/WebKit/WebKit/blob/1b7d7ef4658c992e734bb24c4a3c04c5e4a01ad7/LayoutTests/accessibility/aria-multiline.html is a good example.

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp (hunk 2 above)
Comment (Christop406) [reply Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: Removed this extra element per other comments. Was a remnant of me trying to debug something.

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp (hunk 3 above)
Comment (Christop406) [reply Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: Nope, removed :)

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.h (hunk 9 above)
Comment (Christop406) [reply Source/WebCore/html/shadow/DateTimeNumericFieldElement.h]: Updated the name to be more representative of what it does, per another comment on this PR.

File: Source/WebCore/en.lproj/Localizable.strings (hunk 1 above)
Comment (Christop406) [reply Source/WebCore/en.lproj/Localizable.strings]: Not even sure how i did this lol, removed

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp (hunk 6 above)
Comment (Christop406) [reply Source/WebCore/html/shadow/DateTimeFieldElements.cpp]: yep, both done

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp (hunk 7 above)
Comment (Christop406) [reply Source/WebCore/html/shadow/DateTimeFieldElements.cpp:288]: updated

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp (hunk 11 above)
Comment (Christop406) [reply Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp]: `String::number(value)` didn't work but `AtomString::number(value)` did

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp (hunk 12 above)
Comment (Christop406) [reply Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp]: Good catch!

**Comment by Christop406** [comment]
> Thanks for working on this!
> 
> It would be nice to have a layout test to ensure future changes don't break accessibility. Two cases come to mind:
> 
> 1. Put keyboard focus onto a date field and send up / down keys, and confirm the accessibility label is what we expect
> 2. Update the date input's value via JS, and again verify the accessibility label of the input as a whole and the sub-fields within
> 
> You can read about layout tests here:
> 
> https://github.com/WebKit/WebKit/blob/main/Introduction.md#running-layout-tests
> 
> You'll want to add your new test to `LayoutTests/accessibility`.
> 
> Within the context of layout tests, you can use a global `eventSender` object to simulate events. So for the first testcase, you may want something like this:
> 
> ```
> // Double childAtIndex to get through `<div pseudo="-webkit-datetime-edit">` to `<div pseudo="-webkit-datetime-edit-fields-wrapper">`
> var axDateFieldWrapper = accessibilityController.accessibleElementById("date-input").childAtIndex(0).childAtIndex(0);
> 
> document.getElementById("date-input").focus();
> // Focus should be on the first of the three groups (e.g. the "month" field)
> 
> eventSender.keyDown("upArrow");
> // Verify the description. aria-valuetext is exposed through AccessibilityNodeObject::valueDescription(), which is in turn exposed through AccessibilityUIElement::valueDescription() for layout tests.
> testOutput += expect("axDateFieldWrapper.childAtIndex(0).valueDescription", "AXValueDescription: ...some expected value...");
> 
> // Maybe try some more up / down arrow keys, and verify appropriately
> 
> // Tab to the next field and repeat
> eventSender.keyDown("\t");
> ...
> ```
> 
> In terms of general test structure, https://github.com/WebKit/WebKit/blob/1b7d7ef4658c992e734bb24c4a3c04c5e4a01ad7/LayoutTests/accessibility/aria-multiline.html is a good example.

Just added this. Let me know what you think.

File: Source/WebCore/html/shadow/DateTimeFieldElements.cpp [hunk 13]
```diff
@@ -49,6 +51,8 @@ Ref<DateTimeDayFieldElement> DateTimeDayFieldElement::create(Document& document,
     static MainThreadNeverDestroyed<const AtomString> dayPseudoId("-webkit-datetime-edit-day-field"_s);
     ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
     element->initialize(dayPseudoId);
+    element->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
+    element->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldDayText() });
```
Comment (twilco) [inline Source/WebCore/html/shadow/DateTimeFieldElements.cpp:54]: Seems like this spot still sets aria-label after the role unlike the other two methods.

File: Source/WebCore/html/shadow/DateTimeNumericFieldElement.h [hunk 14]
```diff
@@ -59,6 +59,7 @@ class DateTimeNumericFieldElement : public DateTimeFieldElement {
     void setValueAsInteger(int, EventBehavior = DispatchNoEvent) final;
     void stepDown() final;
     void stepUp() final;
+    void setARIAValueTextWithInteger(int);
```
Comment (twilco) [inline Source/WebCore/html/shadow/DateTimeNumericFieldElement.h]: Can this be private?

File: LayoutTests/accessibility/aria-label-valuetext-date-input.html [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+<head>
+<script src="../resources/accessibility-helper.js"></script>
+<script src="../resources/js-test.js"></script>
+</head>
+<body>
+
+<input id="date-input" type="date" value="2022-02-02" />
+
+<script>
+    var testOutput = "Tests that aria-label & aria-valuetext update correctly for date inputs.\n\n";
```
Comment (twilco) [inline LayoutTests/accessibility/aria-label-valuetext-date-input.html]: Please use "and" instead of & here ([WebKit coding style](https://webkit.org/code-style-guidelines/) prefers to spell things out).

File: LayoutTests/accessibility/aria-label-valuetext-date-input.html [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+    // Getting the wrapper that holds the date inputs.
+    var axDateFieldWrapper = accessibilityController.accessibleElementById("date-input")
+        .childAtIndex(0)
+        .childAtIndex(0);
+
+    function expectChildToHaveDescription(childIdx, descText) {
```
Comment (twilco) [inline LayoutTests/accessibility/aria-label-valuetext-date-input.html]: Please use "childIndex" and "descriptionText" (or "description"), or some other fully spelled out terms.

File: LayoutTests/accessibility/aria-label-valuetext-date-input.html [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+        return expect("axDateFieldWrapper.childAtIndex(" + childIdx + ").valueDescription", "'AXValueDescription: " + descText + "'");
```
Comment (twilco) [inline LayoutTests/accessibility/aria-label-valuetext-date-input.html]: You can use JavaScript template literals here:

```
expect(`axDateFieldWrapper.childAtIndex(${childIndex}).valueDescription`, `"AXValueDescription: ${description}"`);
```

File: LayoutTests/accessibility/aria-label-valuetext-date-input.html [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    function keyDown(key, repeat = 1) {
+        for (let i = 0 ; i < repeat ; i++) {
+            eventSender.keyDown(key);
+        }
```
Comment (twilco) [inline LayoutTests/accessibility/aria-label-valuetext-date-input.html]: The curly braces aren't necessary here.

File: LayoutTests/accessibility/aria-label-valuetext-date-input.html [hunk 19]
```diff
@@ -0,0 +1,84 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
```
Comment (twilco) [inline LayoutTests/accessibility/aria-label-valuetext-date-input.html]: Nice work on the test! Some high-level comments on style:

1. Could we reduce the amount of newlines throughout this test? I personally like to group things by pairing impetus with expectation, so concretely that might look something like:

```
const dateInput = document.getElementById("date-input");
dateInput.focus();
keyDown("upArrow");
testOutput += expectChildToHaveDescription(0, 3);

keyDown("downArrow", 2);
testOutput += expectChildToHaveDescription(0, 1);

keyDown("\t");
testOutput += expectChildToHaveDescription(1, 2);

keyDown("1");
keyDown("0");
testOutput += expectChildToHaveDescription(1, 10);

...
```

2. To make the `-expected.txt` file more readable, it would be nice to log to the test output when you send an event. For example:

```
testOutput += "Simulating up arrow keypress.\n"
keyDown("upArrow");
// ...expectation...
// ...other keypresses / expectations...

testOutput += "Simulating tab key to move to next field in the date input.\n"
keyDown("\t");
...
```

Maybe you could have a helper function:

```
function keyDownAndLog(key) {
    testOutput += `Simulating ${key} keypress.\n`;
    keyDown(key);
}
```

3. Could you make this async? To do that, you'll need to:

Set `window.jsTestIsAsync = true;`, and call `finishJSTest();` after `debug(testOutput);`.

Update `expectChildToHaveDescription` to be like this:

```
async function expectChildToHaveDescription(...args...) {
    return expectAsync(...);
}
```

And because this function is async, you can only call it from async contexts, so I would move most of your test body into an async block like this:

```
setTimeout(async () => {
    ...
});
```

Any calls to `expectChildToHaveDescription` will have to be `await`'d. (Sorry if you know some / all of this).

```
testOutput += await expectChildToHaveDescription(...);
```

Making the test async is important because WebKit accessibility is multi-threaded and eventually consistent, so any dynamic changes won't be immediately reflected. You'll know if you've done it correctly if your test passes with and without the `--accessibility-isolated-tree` flag.

**Comment by Christop406** [comment]
Alright, I updated my PR in response to the second round of review, hopefully it better matches what we're after :).

> (Sorry if you know some / all of this).

@twilco No problem! I do JS development primarily in my job, but I appreciate the specifics here since I want to make it as inline with the other tests as possible!

**Comment by twilco** [comment]
Getting closer! I have a few more comments:

1. Your new test is failing for the GTK port. We'll probably just want to skip the test on GTK, since I think they may not have the necessary flag enabled (`ENABLE(DATE_AND_TIME_INPUT_TYPES)`), or at least their AX structure differs here. You can add:

```
accessibility/aria-label-valuetext-date-input.html [ Skip ]
```

To `LayoutTests/platform/glib/TestExpectations` near where other `accessibility/` tests are skipped.

The test is also failing on Mac WebKit1. It fails there because `<input type="date">` results in this user agent shadow DOM for WK1:

```
<div contenteditable="plaintext-only">2018-07-22</div>
```

So our test isn't relevant. Let's skip it on WK1 too (`LayoutTests/platform/mac-wk1/TestExpectations`).

As an aside, you can run WK1 with the `run-webkit-tests` script by passing the `-1` flag, and you can run WK1 in the browser by using the `Tools/Scripts/run-minibrowser --release` runner, then selecting File > New WebKit1 Window.

2. The third line of your commit message should have the Radar number from Bugzilla (rdar://problem/104928713).

3. I should've caught this earlier, but per [MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/spinbutton_role), `aria-valuenow` is a required attribute for the `spinbutton` role. We're using `aria-valuetext` right now. Could we switch to using `aria-valuenow`? This will unfortunately require changes in the test, too. We expose `aria-valuenow` through `AccessibilityUIElement::intValue()` (so we'll want to replace `valueDescription` with `intValue`, and remove `AXValueDescription: `.

4. You can remove (TODO: Translations?) from your commit message. You've done the right thing with what is in the patch.

I tried your patch out with VoiceOver and it seems like a nice improvement.

File: LayoutTests/accessibility/aria-label-valuetext-date-input.html [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+<head>
+<script src="../resources/accessibility-helper.js"></script>
+<script src="../resources/js-test.js"></script>
+</head>
+<body>
+
+<input id="date-input" type="date" value="2022-02-02" />
+
+<script>
+    var testOutput = "Tests that aria-label and aria-valuetext update correctly for date inputs.\n\n";
```
Comment (twilco) [inline LayoutTests/accessibility/aria-label-valuetext-date-input.html]: I don't think we actually test the aria-label yet 🤔 you should be able to get that with `platformValueForW3CName` (defined in `accessibility-helpers.js`).

What do you think about naming the test something like `date-input-aria-label-and-value.html`? Having date-input up-front reads a bit better IMO, but feel free to tweak that.

Really thorough set of testcases overall, nice job.

**Comment by Christop406** [comment]
@twilco I'm in the process of making your updates, again, all great suggestions! I have a question about the `aria-valuenow` change:

> I should've caught this earlier, but per [MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/spinbutton_role), aria-valuenow is a required attribute for the spinbutton role. We're using aria-valuetext right now. Could we switch to using aria-valuenow? This will unfortunately require changes in the test, too. We expose aria-valuenow through AccessibilityUIElement::intValue() (so we'll want to replace valueDescription with intValue, and remove AXValueDescription: .

If I switch _from_ `aria-valuetext` to `aria-valuenow`, voiceover incorrectly reads the value as "<some number>%" instead of its value. This also happens when setting `aria-valuemin` and `aria-valuemax`. Am I ok to set `aria-valuenow` _and_ `aria-valuetext` to get rid of this issue?

**Comment by twilco** [comment]
> If I switch _from_ `aria-valuetext` to `aria-valuenow`, voiceover incorrectly reads the value as "(some number)%" instead of its value. This also happens when setting `aria-valuemin` and `aria-valuemax`. Am I ok to set `aria-valuenow` _and_ `aria-valuetext` to get rid of this issue?

Ah, good point. Yes, let's set both `aria-valuenow` and `aria-valuetext`.

File: LayoutTests/accessibility/date-input-aria-label-and-value.html [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+                let result = await expectAsync(`axDateFieldWrapper.childAtIndex(${childIndex}).intValue`, `${description}`);
+
+                result += await expectAsync(`axDateFieldWrapper.childAtIndex(${childIndex}).valueDescription`, `'AXValueDescription: ${description}'`);
+
+                return result;
+            }
+            
+            async function expectChildToHaveARIALabel(childIndex, labelText) {
+                return expectAsync(`platformValueForW3CName(axDateFieldWrapper.childAtIndex(${childIndex}))`, labelText);
+            }
```
Comment (twilco) [inline LayoutTests/accessibility/date-input-aria-label-and-value.html]: I don't think this function needs to be async (and it can use `expect` instead of `expectAsync`). The only things that need to be made async in WebKit accessibility layout tests are anything resulting from a dynamic page change via JS or DOM APIs.

File: LayoutTests/accessibility/date-input-aria-label-and-value.html [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+<html>
+<head>
+<script src="../resources/accessibility-helper.js"></script>
+<script src="../resources/js-test.js"></script>
+</head>
+<body>
+
+<input id="date-input" type="date" value="2022-02-02" />
+
+<script>
```
Comment (twilco) [inline LayoutTests/accessibility/date-input-aria-label-and-value.html:11]: Everything inside this `<script>` can be moved back one indentation level.

File: LayoutTests/accessibility/date-input-aria-label-and-value.html [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+
+            keyDown("downArrow", 10, "Simulating down arrow press to decrement the third field value.");
+            testOutput += await expectChildToHaveARIAValueAttributes(2, 2012);
+
+            outputString("Simulating up arrow press to increment the third field value.")
+            keyDown("upArrow");
+            testOutput += await expectChildToHaveARIAValueAttributes(2, 2013);
+
+            outputString("Manually setting the date input's value via JavaScript.")
+            dateInput.value = '2023-12-30';
```
Comment (twilco) [inline LayoutTests/accessibility/date-input-aria-label-and-value.html]: This test mostly uses double quotes, but occasionally uses single quotes (like this spot). It would be nice to use double quotes consistently everywhere.

**Review by twilco — COMMENTED** [review]
Almost there! Just a few style nits left.

I'm not sure why the style checker is failing either as I don't see you referencing the old test name anywhere. Does the style check fail locally? You should be able to use `Tools/Scripts/check-webkit-style`.

**Comment by twilco** [comment]
Ahh, it's failing because you have several different commits at this point. I think you need to squash your changes to a single commit before pushing to Github.

File: LayoutTests/accessibility/date-input-aria-label-and-value.html (hunk 21 above)
Comment (Christop406) [reply LayoutTests/accessibility/date-input-aria-label-and-value.html]: ok good to know for next time (about `expect` vs `expectAsync`)

**Comment by pxlcoder** [comment]
Hi @Christop406, it looks like your commit title is incorrect following the squash, and is missing a description in the message. Would you mind cleaning that up? 

(also, there should be a newline following the rdar link, and before the "Reviewed by" line)

Thank you for your continued iteration on this change!

**Comment by Christop406** [comment]
@pxlcoder Updated this - sorry about that. Thank you guys for continuing to work with me!

**Comment by twilco** [comment]
Oops, your new test is failing for the WPE port. You'll want to skip it there too. You can do that by moving:

```
webkit.org/b/251544 accessibility/date-input-aria-label-and-value.html [ Skip ]
```

from `LayoutTests/platform/gtk/TestExpectations` to `LayoutTests/platform/glib/TestExpectations`

(`glib` encompasses both the GTK and WPE ports).

**Comment by twilco** [comment]
```
This PR adds aria-labels to the inner fields of a date input component.
https://bugs.webkit.org/show_bug.cgi?id=251544
rdar://problem/104928713
Reviewed by NOBODY (OOPS!).
```

I think our bots require a newline between `rdar://problem/104928713` and `Reviewed by NOBODY (OOPS!)` in the commit message, and the top line should probably match the Bugzilla title, so `AX: input[type=date] individual fields are announced as "group"`. Other than that, this is looking great! Almost there 😄

**Comment by twilco** [comment]
Oh, your actual commit message does have the newline, just missing the title update.

**Comment by Christop406** [comment]
> Oh, your actual commit message does have the newline, just missing the title update.

Cool, just updated the commit message, PR title & pushed.

**Comment by twilco** [comment]
Thank you @Christop406! This looks good to me now, but I'm not an official reviewer. @pxlcoder, could you please give this a review and r+ if all looks good?

**Review by pxlcoder — APPROVED** [review]
Thank you for working on this!

If you are interested in another patch, I bet that input[type=time] could benefit from a similar treatment.

**Comment by twilco** [comment]
Thanks for your hard work, this is a great improvement!

> I bet that input[type=time] could benefit from a similar treatment.

This is definitely true. I'd also be happy to offer guidance on any other WebKit accessibility bug that piques your interest. DM me on Slack or start a discussion in any WebKit Slack channel if you have questions.
